### DISCUSSION TITLE
Adding tests for Services

### DISCUSSION THREAD

**Discussion by lloydrichards** [coollabsio/coolify#4290] [Improvement Requests]
There was a comment in #4232 from @Bilge ([here](https://github.com/coollabsio/coolify/pull/4232#issuecomment-2471211012)) that was a bit brutal, but also touches on something important that could be worth discussing:

_How to we make the increasingly large number of services maintainable in the long run through testing?_

I feel this would be a great addition to the [Add a new service template to Coolify](https://coolify.io/docs/contribute/service) documentation on how to properly test the services we're adding and be able to check their continual functionality over time (since many are not version locked).

Does anyone have any suggestions on how we could test this?  And maybe then wrap it up in a github action for CI validation 🤔

**Comment by Bilge** [comment]
Thanks for raising this important discussion. The way I see it, there's two possible paths to take:

1. Test using the API.
2. Test by automating clicks with a UI testing framework like [Selenium](https://www.selenium.dev/) or [Cypress](https://www.cypress.io/).

My preference would be for the former, because I think UI tests can be a little too brittle. Of course, in a perfect world we would have both, but the API approach seems like the best starting point.

**Comment by peaklabs-dev** [comment]
I totally agree that we need testing. I will refactor a lot of code hopefully next week 😅. And then we will add parser test first, which parses the compose files, as that is the most important part. Services tests will probably follow shortly after. Also during the refactoring I will have to delete a lot of services that are super small or not really official, so we will only keep the big and official ones, the other services should not live in the main repo and be added using custom services URL in the future so we keep it the services managable.

**Reply by lloydrichards** [reply]
Awesome to hear this is on the radar! looking forward to seeing how the new service api will work 👀

**Reply by Bilge** [reply]
The [service API](https://coolify.io/docs/api/operations/list-services) isn't new, lol.
